### DISCUSSION TITLE
"How to convert text to path" and "Final drawing not shown until I click the canvas"

### DISCUSSION THREAD

**Discussion by ramoncaballero** [spakin/SimpInkScr#119] [Q&A] [answered]
Thank you for developing this, it's great!

I have (so far) two issues... I'm not sure if I should have created two different discussions...

I'm trying to create a text and apply the equivalent to 'Path > Object to Path' with the following code:

`t = text("10", (100, 100), font_size='16pt').to_path()`

But I get the message: _A path must contain at least one path element._

So I was wondering what is the correct way to convert a text into a path...

What I am actually trying to achieve is this (but with text instead of circles)

```
# Draw a circle:
style(stroke="#ffaaaa", fill="#ffaaaa")
c1 = circle((canvas.width/2, canvas.height/2), 10).to_path()

# Duplicate the 1st circle and send it to the bottom:
style(stroke="#bb8044", fill="#bb8044")
c2 = duplicate(c1)
c2.z_order('bottom', 1)

# Apply the effect 'offset' with 5px to this new circle:
offset5 = path_effect('offset', offset=5, unit='px')
c2.apply_path_effect([offset5])

# Duplicate the 1st circle and send it to the bottom:
style(stroke="#85592c", fill="#85592c")
c3 = duplicate(c1)
c3.z_order('bottom', 1)

# Apply the effect 'offset' with 7px to this new circle:
offset7 = path_effect('offset', offset=7, unit='px')
c3.apply_path_effect([offset7])
```

It works 99% as I would expect. After I run the code above in the Simple Inkscape Scripting dialog (either using 'Python file' or 'Python code' options) only when I click on the 2nd and 3rd circles their final sizes are set as I expect.

If I add `group([c3, c2, c1])` to group the circles, I have to specify the order from bottom to top as `group([c1, c2, c3])` would put c3 on top... is z_order ignored when grouping object?

Thanks in advance.

**Comment by spakin** [comment]
> I have (so far) two issues... I'm not sure if I should have created two different discussions...

Yes, that could help others with the same issue find the discussion.  But I'll address both issues here.

> I'm trying to create a text and apply the equivalent to 'Path > Object to Path' with the following code:
>
> `t = text("10", (100, 100), font_size='16pt').to_path()`
>
> But I get the message: _A path must contain at least one path element._
>
> So I was wondering what is the correct way to convert a text into a path...

The documentation for the underlying `inkex` module, [`inkex.elements._text`](https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._text.html), says,

> Because text is not rendered at all, no information about a text’s path size or actual location can be generated yet.

To elaborate, text objects are fairly opaque to Inkscape extensions.  Only after the extension finishes running does Inkscape render the text, which determines its precise size and location.

A trick you can play is to spawn a child Inkscape that executes *Path* → *Object to Path*.  Here's how:

```python
t = text("10", (100, 100), font_size='16pt')
t = apply_action('object-to-path', t)
```

At some point I'll modify `to_path` to invoke `apply_action` automatically for text objects, but the above should work for now.

> It works 99% as I would expect. After I run the code above in the Simple Inkscape Scripting dialog (either using 'Python file' or 'Python code' options) only when I click on the 2nd and 3rd circles their final sizes are set as I expect.

I'm not able to reproduce this behavior; sorry.  I'm running Inkscape 1.3.2 (091e20e, 2023-11-25) from an AppImage on Ubuntu Linux.

> If I add group([c3, c2, c1]) to group the circles, I have to specify the order from bottom to top as group([c1, c2, c3]) would put c3 on top... is z_order ignored when grouping object?

That's correct.  `group` creates a group then adds each object in turn to the group, putting the new object atop the previously added object.

**Reply by spakin** [reply, marked as answer]
> At some point I'll modify `to_path` to invoke `apply_action` automatically for text objects, but the above should work for now.

I found time to implement this!  (See commit 7c2059c7846881e3324699af4ba34272852f6856.)  Your original
```Python
t = text("10", (100, 100), font_size='16pt').to_path()
```
should now work as expected.

**Reply by ramoncaballero** [reply]
It works, thanks for your answer and for taking your time to implement it.

And thank you for pointing out inkex's documentation, I was not aware of it.

I think I will open a new issue with the second topic I asked here "Final drawing not shown until I click the canvas"

**Reply by spakin** [reply]
As far as grouping goes, I just introduced a [`z_sort`](https://github.com/spakin/SimpInkScr/wiki/Modifying-existing-objects#changing-stacking-order) function that sorts a list of objects by Z-order.  (See commit a4a40dcae48fce3ec2ddaee7ca75a9d4899e40c9.)  In your case, `group(z_sort([c3, c2, c1]))` would add the three shapes to the group while preserving their original Z-order.
